feat: LEAP-1351: LSF customization for bulk annotation

Gondragos opened the pull request:
This PR adds
- `"annotation:bulk"` interface that allows to hide all panels and elements that can't be set equally to group of annotaions
- small fixes for custom controll buttons in bottom bar to be able to use them
- fix the way we deliver config in cypress tests

### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [x] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend


#### What feature flags were used to cover this change?
`fflag_feat_all_leap_1181_bulk_annotation_short`

### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)



### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [x] integration
- [ ] unit



### Which logical domain(s) does this change affect?
`Editor`, `BottomBar`

Gondragos committed:
feat: LEAP-1351: LSF customization for bulk annotation

- hide dependant controls
- hide panels

Gondragos committed:
feat: LEAP-1351: Add ability to customize bottom bar control buttons

Gondragos committed:
tests: LEAP-1351: Add tests

- for bulk annotation mode displaying
- for custom controls rendering and modify on fly
- fix delivering the config from the cypress environment into the test environment

Gondragos committed:
Refactoring

Gondragos committed:
Merge branch 'refs/heads/develop' into fb-leap-1351/bulk-annotation-lsf

Gondragos committed:
biome fixes

yyassi-heartex commented on web/libs/editor/src/components/App/App.jsx:
```diff
@@ -260,16 +268,23 @@ class App extends Component {
-                  {mainContent}
-                  {store.hasInterface("topbar") && <BottomBar store={store} />}
-                </SideTabsPanels>
+                shouldShowPanels ? (
```
should this be applied for situations where `newUIEnabled` is false? IE are we supporting this for situations where FF_DEV_3873 is turned off?

Gondragos replied:
Thank you! I have checked we have this flag switched on not everywhere.
So I've changed the code and now it might be displayed right for both cases if `FF_DEV_1170` is on.

yyassi-heartex commented on web/libs/editor/src/tags/visual/Collapse.jsx:
```diff
@@ -23,76 +24,107 @@ const { Panel } = Collapse;
-        .filter((i) => i.type === "panel")
+        .filter(
+          (i) =>
+            i.type === "panel" &&
+            ((!isFF(FF_BULK_ANNOTATION) && !item.store.hasInterface("bulk:annotation")) || i.isIndependent),
```
can we move `(!isFF(FF_BULK_ANNOTATION) && !item.store.hasInterface("bulk:annotation")` to a variable outside the filter that describes what this means? - this way we dont need to recalculate it for every `i` when 'i' doesn't affect it.
this will also help better define what this is making it more readable

Gondragos replied:
Now I think that the condition was also wrong.

yyassi-heartex commented on web/libs/editor/src/tags/visual/Filter.jsx:
```diff
@@ -59,6 +59,11 @@ const Model = types
+
+    // Indicates that it could exist without information about objects, taskData and regions
+    get isIndependent() {
+      return self.toTag ? self.toTag.isIndependent : false;
```
same as this no?
```suggestion
      return self.toTag && self.toTag.isIndependent;
```

hlomzik replied:
@yyassi-heartex your version returns mixed values instead of plain boolean, so I'd prefer the original version

yyassi-heartex replied:
ahh i see - good point

Gondragos replied:
@hlomzik From the other side, for some tags it can be `undefined` anyhow.

yyassi-heartex replied:
since its really just checking if toTag is defined maybe we can do something like - feels like thats the intent - is that spot on?
```suggestion
      return self.toTag?.isIndependent ?? false;
```

Gondragos committed:
Refactoring

hlomzik commented:
Would not it be better to name interface "annotation:bulk"? That will keep it in a shape of namespace:detail like for lot of other interfaces. I know they are a mess but that's a really lowcost change on an early stage with small benefit

Gondragos committed:
Rename interface

Gondragos committed:
Hide unnecessary elements for both modes with outliner

hlomzik commented on web/libs/editor/src/components/App/App.jsx:
```diff
@@ -219,6 +226,7 @@ class App extends Component {
       </Block>
     );
 
+    const shouldShowPanels = isDefined(store) && (!isFF(FF_BULK_ANNOTATION) || !store.hasInterface("annotation:bulk"));
```
`store` is definitely defined at this point. and I'd suggest to rename it to `isBulkAnnotation` so later in the code it will be more clear when we display panels and when we don't

Gondragos replied:
It kinda depends on what you want to make clear, but I'm ok with that change.

Gondragos committed:
Rename variable

Gondragos committed:
Fix misprint

Gondragos committed:
Hide more things and update tests

hlomzik commented on web/libs/editor/src/core/CustomTypes.ts:
```diff
@@ -50,7 +50,30 @@ const CSSColor = types.custom<any, string>({
+// and be able to get a reaction on its change
+//
+// /!\ Avoid using this type in case you need serialization of the model it contains
+const RawCallback = types.custom<Function, Function>({
```
I think we definitely need to add more explanations on how and why would we use this new type, including:
- what was the high-level reason of adding new uncommon type to MST
- how would we create modals with this types from outside, examples maybe?
- what other considerations were there and why this one is the best option

Why I think that these are important questions they can be answered later.

Gondragos replied:
It was decided to take the solution of this issue beyond this PR

hlomzik commented on web/libs/editor/src/tags/visual/View.jsx:
It's better to finally move this list of allowed children from this and other tags to a reasonable place (like in tags/children or some constants list), plus optionally add an ability to easily modify them if we don't accept one specific tag for example. But that's for later.

hlomzik commented on web/libs/editor/src/core/Tree.tsx:
```diff
@@ -234,6 +234,10 @@ function renderItem(ref: IAnyStateTreeNode, annotation: IAnnotation, includeKey
   const typeName = type.name;
   const View = Registry.getViewByModel(typeName);
 
+  if (isFF(FF_BULK_ANNOTATION) && annotation?.store?.hasInterface("annotation:bulk") && el.isIndependent !== true) {
```
I'd split this too long line into `isBulk` const and short check

hlomzik commented on web/libs/editor/src/components/BottomBar/BottomBar.jsx:
```diff
@@ -20,7 +20,11 @@ export const BottomBar = observer(({ store }) => {
         {store.hasInterface("controls") && (store.hasInterface("review") || !isPrediction) && (
           <Elem name="section" mod={{ flat: true }}>
-            <Controls annotation={entity} />
+            {isFF(FF_BULK_ANNOTATION) && store.hasInterface("controls:custom") ? (
```
Add this interface to the PR description

Gondragos committed:
Refactor condition

Separated bulk mode and independence checks into variables for readability. This makes the conditional logic clearer and potentially reduces redundancy in code review and debugging processes. No functional changes have been introduced.

Gondragos committed:
Merge branch 'develop' into fb-leap-1351/bulk-annotation-lsf

# Conflicts:
#	web/libs/editor/src/components/BottomBar/BottomBar.jsx

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10562522580)
> Successfully merged: Already up to date.

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10617538191)
> Successfully merged:  387 files changed, 6483 insertions(+), 9039 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1351/bulk-annotation-lsf'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10617538191

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10674028201)
> Successfully merged:  29 files changed, 94 insertions(+), 69 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1351/bulk-annotation-lsf'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10674028201

robot-ci-heartex commented:
This PR is stale because it has been open 45 days with no activity. Remove `stale` label or comment or this will be closed in 10 days.

robot-ci-heartex commented:
This PR was closed because it has been stalled for 10 days with no activity.

Gondragos committed:
Merge branch 'develop' into fb-leap-1351/bulk-annotation-lsf

# Conflicts:
#	web/libs/editor/src/components/App/App.jsx
#	web/libs/editor/src/components/BottomBar/Controls.tsx
#	web/libs/editor/src/stores/AppStore.js
#	web/libs/editor/src/tags/control/ClassificationBase.js
#	web/libs/editor/src/tags/visual/Collapse.jsx
#	web/libs/editor/src/tags/visual/View.jsx

Gondragos committed:
Adjust types, small fixes, add ability to update button

Gondragos committed:
Adjust tests

Gondragos committed:
Merge branch 'develop' into fb-leap-1351/bulk-annotation-lsf

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11975726292)
> Successfully merged:  29 files changed, 78 insertions(+), 117 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1351/bulk-annotation-lsf'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/11975726292

Gondragos committed:
Remove unused parts of code

Gondragos committed:
Change way of applying state to CustomButton

Gondragos committed:
Refactoring

Gondragos committed:
Merge branch 'fb-leap-1351/bulk-annotation-lsf' of ssh://github.com/heartexlabs/label-studio into fb-leap-1351/bulk-annotation-lsf

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12011593003)
> Successfully merged:  12 files changed, 21 insertions(+), 40 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1351/bulk-annotation-lsf'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/12011593003

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/12033118881)
> Successfully merged:  delete mode 100644 Dockerfile.redhat

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1351/bulk-annotation-lsf'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/12033118881

hlomzik commented on web/libs/editor/src/utils/feature-flags.ts:
```diff
@@ -272,6 +272,10 @@ export const FF_REVIEWER_FLOW = "fflag_feat_all_leap_1081_reviewer_flow_updates"
+/**
+ * It adds functionality of bulk annotation
+ */
+export const FF_BULK_ANNOTATION = "fflag_feat_all_leap_1181_bulk_annotation_short";
```
```suggestion
export const FF_BULK_ANNOTATION = "fflag_feat_all_leap_1181_bulk_annotation_short";

```

hlomzik commented on web/libs/editor/src/tags/visual/Collapse.jsx:
```diff
@@ -22,76 +23,105 @@ const { Panel } = Collapse;
 
-  _value: types.optional(types.string, ""),
-  value: types.optional(types.string, ""),
+      // In other cases Panel can exist independent if it hase some independent children
```
```suggestion
      // In other cases Panel can exist independent if it has some independent children
```

Commit:
Apply suggestions from code review

Co-authored-by: hlomzik <hlomzik@gmail.com>

yyassi-heartex reviewed (approved):
looks good, but as @nicholasrq pointed out unit tests need to be fixed - approving with the caveat that they be fixed

nick-skriabin commented on web/libs/editor/src/components/BottomBar/buttons.tsx:
```diff
@@ -15,15 +15,15 @@ type MixedInParams = {
 
-export function controlsInjector<T extends {}>(fn: (props: T & MixedInParams) => ReactNode) {
+export function controlsInjector<T extends {}>(fn: (props: T & MixedInParams) => ReactElement) {
   const wrapped = inject(({ store }) => {
     return {
       store,
       history: store?.annotationStore?.selected?.history,
     };
   })(fn);
   // inject type doesn't handle the injected props, so we have to force cast it
-  return wrapped as unknown as (props: T) => ReactNode;
+  return wrapped as unknown as (props: T) => ReactElement;
```
you don't need to specify return type in the function signature if you explicitly do that in the return statement

Gondragos replied:
At first a thought that you are right but then I figured out that it's different variables.
<img width="873" alt="image" src="https://github.com/user-attachments/assets/85fce9e9-048d-460e-b23a-8c0ae899a951">

nick-skriabin commented on web/libs/editor/src/components/App/App.jsx:
```diff
@@ -278,24 +280,33 @@ class App extends Component {
-                  {store.hasInterface("topbar") && <BottomBar store={store} />}
-                </SideTabsPanels>
-              ) : (
+                !isBulkMode ? (
+                  <SideTabsPanels
+                    panelsHidden={viewingAll}
+                    currentEntity={as.selectedHistory ?? as.selected}
+                    regions={as.selected.regionStore}
+                    showComments={store.hasInterface("annotations:comments")}
+                    focusTab={store.commentStore.tooltipMessage ? "comments" : null}
+                  >
+                    {mainContent}
+                    {store.hasInterface("topbar") && <BottomBar store={store} />}
+                  </SideTabsPanels>
+                ) : (
+                  <>
+                    {mainContent}
+                    {store.hasInterface("topbar") && <BottomBar store={store} />}
+                  </>
+                )
```
i would reverse this ternary so the bulk UI goes first. `!isBulk` type of notation are harder to read in such case

yyassi-heartex replied:
thank you for pointing this out - im considering adding this lint rule https://biomejs.dev/linter/rules/no-negation-else/#_top to help catch this with the linter

bmartel commented on web/libs/editor/src/components/TopBar/TopBar.jsx:
```diff
@@ -77,8 +80,8 @@ export const TopBar = observer(({ store }) => {
           <Elem name="group">
-            <CurrentTask store={store} />
-            {!isViewAll && (
+            {!isBulkMode && <CurrentTask store={store} />}
+            {!isViewAll && !isBulkMode && (
```
Is this even possible to reach here with `isBulkMode` true given the check at the top of the component returns null in that case? And if it cannot be true, it is presumed only false here.

Gondragos replied:
Yep, we can have feature flag `FF_DEV_3873` turned off.
I mean it shouldn't be a usual case but for some reason I thought it's better to cover.
Although i'm not sure right now we cover all other functionality with `FF_DEV_3873` turned off, so maybe it's better to remove these 2 conditions at all. 🤔

Gondragos replied:
I have checked that. This conditions are covered by the test https://github.com/HumanSignal/label-studio/pull/6203/files#diff-15e1d28fc5b0c407a694fe6996ee68a73faa93b0ea1aa5d47bec1dd585ef82c8R146-R158
So I would say it's better to keep them there untill we remove the feature flag about old ui.

bmartel commented on web/libs/editor/src/tags/visual/Collapse.jsx:
```diff
@@ -22,76 +23,105 @@ const { Panel } = Collapse;
+    get isIndependent() {
+      // if value starts with $ it's related to the data but in case of Panel it's just affect title
+      // we may still want to show it even if there is no data
+      // if (self.value && self.value[0] === "$") return false;
```
```
      // if (self.value && self.value[0] === "$") return false;
```

Is this meant to be commented out? I would remove this if unused.

Gondragos replied:
Yeah... At first I thought that it was a good idea to hide this element of it depends on a data field. But then I figured out that it could be better to display it anyway. This line is  "ready to be uncomented" in case if my last assumption is not right. But I would like to keep it just as an option for a while.

Maybe there is a better way to describe it in the comment but I didn't find it yet. 😅

bmartel reviewed (approved):
Couple tiny little things, really well done!

Gondragos committed:
Refactoring
